### PR TITLE
Install middleware in ObjectAdapter in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR improves ObjectAdapter in C# to allow users to install middleware and construct the dispatch pipeline for the ObjectAdapter.

It also a new middleware test that currently only verifies middleware execution order.

### CODE REVIEW

File: csharp/src/Ice/Internal/ServantManager.cs [hunk 1]
```diff
@@ -6,26 +6,91 @@
 
 namespace Ice.Internal;
 
-public sealed class ServantManager : Object
+internal sealed class ServantManager : Object
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/ServantManager.cs:9]: I actually did not really change this file, just cleaned it up.

File: csharp/src/Ice/ObjectAdapter.cs [hunk 2]
```diff
@@ -382,6 +385,24 @@ public void destroy()
...
+    /// <param name="middleware">The middleware to install.</param>
+    /// <returns>This object adapter.</returns>
+    /// <exception cref="InvalidOperationException">Thrown if the object adapter's dispatch pipeline has already been
+    /// created. This creation typically occurs the first time the object adapter dispatches an incoming request.
+    /// </exception>
+    public ObjectAdapter use(Func<Object, Object> middleware)
+    {
+        if (_dispatchPipeline.IsValueCreated)
+        {
+            throw new InvalidOperationException("All middleware must be installed before the first dispatch.");
```
[the remark this answers is not in the document]
Comment (externl) [reply csharp/src/Ice/ObjectAdapter.cs:400]: I'd rather not created a new LocalException if the language provides a reasonable alternative. We'll have no choice but to create or reuse one in some languages, like Swift.

**Review by externl — APPROVED** [review]

File: csharp/test/Ice/metrics/Server.cs [hunk 3]
```diff
@@ -28,7 +28,7 @@ public override void run(string[] args)
 
         communicator.getProperties().setProperty("ForwardingAdapter.Endpoints", getTestEndpoint(1));
         Ice.ObjectAdapter forwardingAdapter = communicator.createObjectAdapter("ForwardingAdapter");
-        forwardingAdapter.addDefaultServant(adapter.dispatcher, "");
+        forwardingAdapter.addDefaultServant(adapter.dispatchPipeline, "");
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/test/Ice/metrics/Server.cs:31]: Why do you think it should fail?

File: csharp/test/Ice/middleware/AllTests.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+#nullable enable
+
+namespace Ice.middleware;
+
+public class AllTests : global::Test.AllTests
+{
+    public static void allTests(global::Test.TestHelper helper)
+    {
+        var output = helper.getWriter();
+        Ice.Communicator communicator = helper.communicator();
```
Comment (pepone) [inline csharp/test/Ice/middleware/AllTests.cs:12]:
```suggestion
       Communicator communicator = helper.communicator();
```

File: csharp/test/Ice/middleware/AllTests.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        output.Write("testing middleware execution order... ");
+        output.Flush();
+
+        // Arrange
+        var inLog = new List<string>();
+        var outLog = new List<string>();
+
+        var oa = communicator.createObjectAdapter("");
+
+        oa.add(new MyObjectI(), Ice.Util.stringToIdentity("test"));
```
Comment (pepone) [inline csharp/test/Ice/middleware/AllTests.cs:28]: Seems simpler to use `new Identity { name = "test" }` in new code.

**Review by pepone — APPROVED** [review]

File: csharp/test/Ice/metrics/Server.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/test/Ice/metrics/Server.cs:31]: With this update, the metrics middleware runs twice when using the forwarding adapter. Before, it was running only once. I thought this double-collection of metrics would make the test fail, but it didn't.

File: csharp/test/Ice/middleware/AllTests.cs (hunk 5 above)
Comment (bernardnormier) [reply csharp/test/Ice/middleware/AllTests.cs:28]: fixed

**Pull request merged**